linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala

```diff
@@ -0,0 +1,392 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.backend.webassembly
+
+import scala.collection.mutable
+
+import org.scalajs.ir.{OriginalName, Position}
+
+import Instructions._
+import Identitities._
+import Modules._
+import Types._
+
+final class FunctionBuilder(
+    moduleBuilder: ModuleBuilder,
+    val functionID: FunctionID,
+    val functionOriginalName: OriginalName,
+    functionPos: Position
+) {
+  import FunctionBuilder._
+
+  private var labelIdx = 0
+
+  private val params = mutable.ListBuffer.empty[Local]
+  private val locals = mutable.ListBuffer.empty[Local]
+  private var resultTypes: List[Type] = Nil
+
+  private var specialFunctionType: Option[TypeID] = None
+
+  /** The instructions buffer. */
+  private val instrs: mutable.ListBuffer[Instr] = mutable.ListBuffer.empty
+
+  def setFunctionType(typeID: TypeID): Unit =
+    specialFunctionType = Some(typeID)
+
+  def setResultTypes(tpes: List[Type]): Unit =
+    resultTypes = tpes
+
+  def setResultType(tpe: Type): Unit =
+    setResultTypes(tpe :: Nil)
```

Comment:
They're not mutually exclusive. I added a longer comment on `setFunctionType` to clarify its purpose, and also to clarify why it is not exclusive with `setResultType(s)`.

Given that they're not mutually exclusive, adding them to constructor parameters would directly lead to some combinatorial explosion. The purpose of a `Builder` object is to avoid this kind of issue by admitting some mutability.